File: OsmAndWear/src/net/osmand/plus/LockableViewPager.java
package net.osmand.plus;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.viewpager.widget.ViewPager;

public class LockableViewPager extends ViewPager {
	private static final int POS_Y_UNLOCKED = -1;
	private static final int POS_Y_LOCKED = Integer.MAX_VALUE;
	private int swipeLockedPosY = POS_Y_UNLOCKED;

	public LockableViewPager(Context context) {
		super(context);
	}

	public LockableViewPager(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public boolean getSwipeLocked() {
		return swipeLockedPosY != POS_Y_UNLOCKED;
	}

	public void setSwipeLocked(boolean swipeLocked) {
		this.swipeLockedPosY = swipeLocked ? POS_Y_LOCKED : POS_Y_UNLOCKED;
	}

	public void setSwipeLockedPosY(int posY) {
		this.swipeLockedPosY = posY;
	}

	@Override
	public boolean onTouchEvent(MotionEvent event) {
		return (swipeLockedPosY == POS_Y_UNLOCKED || swipeLockedPosY < event.getY()) && super.onTouchEvent(event);
	}

	@Override
	public boolean onInterceptTouchEvent(MotionEvent event) {
		return (swipeLockedPosY == POS_Y_UNLOCKED || swipeLockedPosY < event.getY()) && super.onInterceptTouchEvent(event);
	}

	@Override
	public boolean canScrollHorizontally(int direction) {
		return swipeLockedPosY != POS_Y_LOCKED && super.canScrollHorizontally(direction);
	}
}

File: OsmAnd-telegram/src/net/osmand/telegram/ui/views/LockableViewPager.kt
package net.osmand.telegram.ui.views

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent

class LockableViewPager : androidx.viewpager.widget.ViewPager {
	var swipeLocked: Boolean = false

	constructor(context: Context) : super(context) {}

	constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

	override fun onTouchEvent(event: MotionEvent): Boolean {
		return !swipeLocked && super.onTouchEvent(event)
	}

	override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
		return !swipeLocked && super.onInterceptTouchEvent(event)
	}

	override fun canScrollHorizontally(direction: Int): Boolean {
		return !swipeLocked && super.canScrollHorizontally(direction)
	}
}
